File: packages/guards/src/accessibility-output.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface AccessibilityOutputOptions { action: 'block' | 'warn'; maxReadingLevel?: number; }

function countSyllables(word: string): number {
  const w = word.toLowerCase().replace(/(?:[^laeiouy]es|ed|[^laeiouy]e)$/, '').replace(/^y/, '');
  const vowelGroups = w.match(/[aeiouy]{1,2}/g);
  return vowelGroups ? Math.max(vowelGroups.length, 1) : 1;
}

function estimateGradeLevel(text: string): number {
  const sentences = text.split(/[.!?]+/).filter(s => s.trim().length > 0);
  if (sentences.length === 0) return 0;
  const words = text.split(/\s+/).filter(w => /[a-zA-Z]/.test(w));
  if (words.length === 0) return 0;
  const totalSyllables = words.reduce((sum, w) => sum + countSyllables(w), 0);
  const avgWordsPerSentence = words.length / sentences.length;
  const avgSyllablesPerWord = totalSyllables / words.length;
  const grade = 0.39 * avgWordsPerSentence + 11.8 * avgSyllablesPerWord - 15.59;
  return Math.max(0, Math.round(grade * 10) / 10);
}

function checkImageAltText(text: string): string[] {
  const issues: string[] = [];
  const imgRegex = /!\[([^\]]*)\]\([^)]+\)/g;
  let m: RegExpExecArray | null;
  while ((m = imgRegex.exec(text)) !== null) {
    const alt = m[1].trim();
    if (!alt) { issues.push('missing_alt_text'); }
    else if (alt.length < 5 || /^(?:image|photo|picture|img|screenshot)$/i.test(alt)) { issues.push('non_descriptive_alt_text'); }
  }
  return issues;
}

function checkColorOnly(text: string): boolean {
  return /\b(?:the\s+)?(?:red|green|blue|yellow|orange|purple|pink)\s+(?:items?|buttons?|sections?|areas?|text|elements?|boxes?|markers?|indicators?)\b/i.test(text)
    && !/\b(?:labeled|marked|titled|numbered|named|with\s+(?:icon|symbol|label|text))\b/i.test(text);
}

function checkLinkText(text: string): string[] {
  const issues: string[] = [];
  const linkRegex = /\[([^\]]+)\]\([^)]+\)/g;
  let m: RegExpExecArray | null;
  while ((m = linkRegex.exec(text)) !== null) {
    const linkText = m[1].trim().toLowerCase();
    if (/^(?:click\s+here|here|read\s+more|more|link|this)$/i.test(linkText)) { issues.push(`inaccessible_link_text: "${m[1]}"`); }
  }
  return issues;
}

function checkAbbreviations(text: string): string[] {
  const issues: string[] = [];
  const abbrRegex = /\b([A-Z]{2,6})\b/g;
  const defined: Record<string, boolean> = {};
  const parenDef = /\(([A-Z]{2,6})\)/g;
  let m: RegExpExecArray | null;
  while ((m = parenDef.exec(text)) !== null) defined[m[1]] = true;
  const colonDef = /([A-Z]{2,6})\s*[-–:]\s*[A-Z][a-z]/g;
  while ((m = colonDef.exec(text)) !== null) defined[m[1]] = true;
  const common: Record<string, boolean> = { US: true, UK: true, EU: true, AI: true, IT: true, OK: true, AM: true, PM: true, ID: true, TV: true, PC: true, OS: true, VS: true, OR: true, AN: true, IF: true, IS: true, IN: true, ON: true, TO: true, DO: true, GO: true, NO: true, SO: true, UP: true, AT: true, BY: true, AS: true, OF: true };
  while ((m = abbrRegex.exec(text)) !== null) {
    if (!defined[m[1]] && !common[m[1]]) { issues.push(`undefined_abbreviation: ${m[1]}`); defined[m[1]] = true; }
  }
  return issues;
}

function checkTableHeaders(text: string): boolean {
  const tableRegex = /\|[^\n]+\|\n/;
  const headerSep = /\|[\s:]*-+[\s:]*\|/;
  if (tableRegex.test(text) && !headerSep.test(text)) return true;
  return false;
}

export function accessibilityOutput(options: AccessibilityOutputOptions): Guard {
  const maxGrade = options.maxReadingLevel ?? 12;
  return { name: 'accessibility-output', version: '0.1.0', description: 'Verify output meets accessibility standards (WCAG 2.2 inspired)', category: 'content', supportedStages: ['output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now(); const issues: string[] = [];
      const grade = estimateGradeLevel(text);
      if (grade > maxGrade) issues.push(`reading_level_too_high: grade ${grade} > ${maxGrade}`);
      issues.push(...checkImageAltText(text));
      if (checkColorOnly(text)) issues.push('color_only_information');
      issues.push(...checkLinkText(text));
      issues.push(...checkAbbreviations(text));
      if (checkTableHeaders(text)) issues.push('table_missing_header_row');
      const triggered = issues.length > 0;
      return { guardName: 'accessibility-output', passed: !triggered, action: triggered ? options.action : 'allow',
        message: triggered ? `Accessibility issues found: ${issues.length}` : undefined,
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { issues, gradeLevel: grade, maxReadingLevel: maxGrade } : undefined,
      };
    },
  };
}

File: python/open_guardrail/guards/accessibility_output.py
"""Verify output meets accessibility standards (WCAG 2.2 inspired)."""
from __future__ import annotations

import re
import time
from typing import List, Optional

from open_guardrail.core import GuardResult


def _count_syllables(word: str) -> int:
    w = word.lower()
    w = re.sub(r"(?:[^laeiouy]es|ed|[^laeiouy]e)$", "", w)
    w = re.sub(r"^y", "", w)
    groups = re.findall(r"[aeiouy]{1,2}", w)
    return max(len(groups), 1) if groups else 1


def _estimate_grade(text: str) -> float:
    sentences = [
        s
        for s in re.split(r"[.!?]+", text)
        if s.strip()
    ]
    if not sentences:
        return 0.0
    words = [
        w for w in text.split() if re.search(r"[a-zA-Z]", w)
    ]
    if not words:
        return 0.0
    total_syl = sum(_count_syllables(w) for w in words)
    avg_wps = len(words) / len(sentences)
    avg_spw = total_syl / len(words)
    grade = 0.39 * avg_wps + 11.8 * avg_spw - 15.59
    return round(max(0.0, grade), 1)


def _check_image_alt(text: str) -> List[str]:
    issues: List[str] = []
    for m in re.finditer(r"!\[([^\]]*)\]\([^)]+\)", text):
        alt = m.group(1).strip()
        if not alt:
            issues.append("missing_alt_text")
        elif len(alt) < 5 or re.match(
            r"^(?:image|photo|picture|img|screenshot)$",
            alt,
            re.I,
        ):
            issues.append("non_descriptive_alt_text")
    return issues


def _check_color_only(text: str) -> bool:
    has_color = bool(
        re.search(
            r"\b(?:the\s+)?(?:red|green|blue|yellow|"
            r"orange|purple|pink)\s+"
            r"(?:items?|buttons?|sections?|areas?|"
            r"text|elements?|boxes?|markers?|"
            r"indicators?)\b",
            text,
            re.I,
        )
    )
    has_alt = bool(
        re.search(
            r"\b(?:labeled|marked|titled|numbered|"
            r"named|with\s+(?:icon|symbol|label|text))\b",
            text,
            re.I,
        )
    )
    return has_color and not has_alt


def _check_link_text(text: str) -> List[str]:
    issues: List[str] = []
    for m in re.finditer(r"\[([^\]]+)\]\([^)]+\)", text):
        lt = m.group(1).strip().lower()
        if re.match(
            r"^(?:click\s+here|here|read\s+more|"
            r"more|link|this)$",
            lt,
            re.I,
        ):
            issues.append(
                f'inaccessible_link_text: "{m.group(1)}"'
            )
    return issues


def _check_abbreviations(text: str) -> List[str]:
    issues: List[str] = []
    defined = set()
    for m in re.finditer(r"\(([A-Z]{2,6})\)", text):
        defined.add(m.group(1))
    for m in re.finditer(
        r"([A-Z]{2,6})\s*[-\u2013:]\s*[A-Z][a-z]", text
    ):
        defined.add(m.group(1))
    common = {
        "US", "UK", "EU", "AI", "IT", "OK",
        "AM", "PM", "ID", "TV", "PC", "OS",
        "VS", "OR", "AN", "IF", "IS", "IN",
        "ON", "TO", "DO", "GO", "NO", "SO",
        "UP", "AT", "BY", "AS", "OF",
    }
    for m in re.finditer(r"\b([A-Z]{2,6})\b", text):
        abbr = m.group(1)
        if abbr not in defined and abbr not in common:
            issues.append(
                f"undefined_abbreviation: {abbr}"
            )
            defined.add(abbr)
    return issues


def _check_table_headers(text: str) -> bool:
    has_table = bool(re.search(r"\|[^\n]+\|\n", text))
    has_sep = bool(
        re.search(r"\|[\s:]*-+[\s:]*\|", text)
    )
    return has_table and not has_sep


class _AccessibilityOutput:
    def __init__(
        self,
        *,
        action: str = "warn",
        max_reading_level: int = 12,
    ) -> None:
        self.name = "accessibility-output"
        self.action = action
        self.max_reading_level = max_reading_level

    def check(
        self, text: str, stage: str = "output"
    ) -> GuardResult:
        start = time.perf_counter()
        issues: List[str] = []

        grade = _estimate_grade(text)
        if grade > self.max_reading_level:
            issues.append(
                f"reading_level_too_high:"
                f" grade {grade}"
                f" > {self.max_reading_level}"
            )

        issues.extend(_check_image_alt(text))
        if _check_color_only(text):
            issues.append("color_only_information")
        issues.extend(_check_link_text(text))
        issues.extend(_check_abbreviations(text))
        if _check_table_headers(text):
            issues.append("table_missing_header_row")

        triggered = len(issues) > 0
        elapsed = (time.perf_counter() - start) * 1000

        return GuardResult(
            guard_name="accessibility-output",
            passed=not triggered,
            action=(
                self.action if triggered else "allow"
            ),
            message=(
                f"Accessibility issues found:"
                f" {len(issues)}"
                if triggered
                else None
            ),
            latency_ms=round(elapsed, 2),
            details=(
                {
                    "issues": issues,
                    "grade_level": grade,
                    "max_reading_level": (
                        self.max_reading_level
                    ),
                }
                if triggered
                else None
            ),
        )


def accessibility_output(
    *,
    action: str = "warn",
    max_reading_level: int = 12,
) -> _AccessibilityOutput:
    return _AccessibilityOutput(
        action=action,
        max_reading_level=max_reading_level,
    )
